x-pack/plugins/siem/server/lib/machine_learning/authz.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+
+import { KibanaRequest } from '../../../../../../src/core/server/';
+import { ILicense } from '../../../../licensing/server';
+import { MlPluginSetup } from '../../../../ml/server';
+import { SetupPlugins } from '../../plugin';
+import { MINIMUM_ML_LICENSE } from '../../../common/constants';
+import { hasMlAdminPermissions } from '../../../common/machine_learning/has_ml_admin_permissions';
+import { isMlRule } from '../../../common/machine_learning/helpers';
+import { RuleType } from '../../../common/detection_engine/types';
+import { Validation } from './validation';
+
+export interface MlAuthz {
+  validateRuleType: (type: RuleType) => Promise<Validation>;
+}
+
+/**
+ * Builds ML authz services
+ *
+ * @param license A {@link ILicense} representing the user license
+ * @param ml {@link MlPluginSetup} ML services to fetch ML capabilities
+ * @param request A {@link KibanaRequest} representing the authenticated user
+ *
+ * @returns A {@link MLAuthz} service object
+ */
+export const buildMlAuthz = ({
+  license,
+  ml,
+  request,
+}: {
+  license: ILicense;
+  ml: SetupPlugins['ml'];
+  request: KibanaRequest;
+}): MlAuthz => {
+  let validation: Promise<Validation> | null = null;
```

Comment:
Woa...Does this do caching? If so, and the validation fails even once for a different user and the first user does that mean all users are going to get that same cache result?

Caches within the backend are going to be dangerous with requests. We don't do this anywhere else at the moment do we?

Reply:
@FrankHassanabad the value of the returned function is cached, correct. I don't know of us doing this elsewhere. The lifecycle of this function is a single request (and you have to provide a request to receive a function), so I was thinking we were okay here, but I'm open to suggestions.

I did this to satisfy a few requirements:

1. prevent calling ML services if the request isn't doing ML stuff
    * satisfied by lazy evaluation
2. If invoking in a loop (import, bulk create, bulk update, etc.), prevent calling ML services on each invocation
    * satisfied with caching
    * this is because in these rule routes, the authorization call is not dependent on a particular rule (just request-level information), but whether we _care_ about the result of that authorization call is.

Reply:
Thanks for the explanation. Sounds good! 👍 , I'm just always careful around server side caches.

Reply:
Two tests if you want them. I wrote them to make sure I understood what was going on. Everything looks good like you can call the builder twice or just once and it captures and holds the result as you want it. I like the pattern!  👍 

```ts
    it('validates the same cache result per request if permissions change mid-stream', async () => {
      licenseMock.hasAtLeast.mockReturnValueOnce(false);
      licenseMock.hasAtLeast.mockReturnValueOnce(true);

      const mlAuthz = buildMlAuthz({
        license: licenseMock,
        ml: mlMock,
        request,
      });

      const validationFirst = await mlAuthz.validateRuleType('machine_learning');
      const validationSecond = await mlAuthz.validateRuleType('machine_learning');

      expect(validationFirst.valid).toEqual(false);
      expect(validationFirst.message).toEqual(
        'Your license does not support machine learning. Please upgrade your license.'
      );
      expect(validationSecond.valid).toEqual(false);
      expect(validationSecond.message).toEqual(
        'Your license does not support machine learning. Please upgrade your license.'
      );
    });

    it('will invalidate the cache result if the builder is called a second time after a license change', async () => {
      licenseMock.hasAtLeast.mockReturnValueOnce(false);
      licenseMock.hasAtLeast.mockReturnValueOnce(true);
      (hasMlAdminPermissions as jest.Mock).mockReturnValueOnce(true);

      const mlAuthzFirst = buildMlAuthz({
        license: licenseMock,
        ml: mlMock,
        request,
      });

      const mlAuthzSecond = buildMlAuthz({
        license: licenseMock,
        ml: mlMock,
        request,
      });

      const validationFirst = await mlAuthzFirst.validateRuleType('machine_learning');
      const validationSecond = await mlAuthzSecond.validateRuleType('machine_learning');

      expect(validationFirst.valid).toEqual(false);
      expect(validationFirst.message).toEqual(
        'Your license does not support machine learning. Please upgrade your license.'
      );
      expect(validationSecond.valid).toEqual(true);
      expect(validationSecond.message).toBeUndefined();
    });
```